### PR TITLE
Feature reworked install process

### PR DESCRIPTION
**Pull request by bvanessen**
Updated the spack build scripts and instructions to use the `spack dev-build` and `spack build-env` commands.  Also makes it simpler to find dependencies using `spack externals find`.

### CODE REVIEW

**Comment by mrwyattii** [comment]
From trying the new install methods on Pascal...

USER EXPERIENCE:
I could not get a non-developer install to work with the instructions provided in the docs ('./scripts/build_lbann.sh -e lbann`), giving the error: `==> Error: invalid values for variant "cuda_arch" in package "lbann": ['60+dihydrogen+distconv~half']`

DEVELOPER EXPERIENCE:
Using the developer install instructions (`/scripts/build_lbann.sh -d -- +dihydrogen +cuda +half`) I was able to successfully build LBANN.

The docs and instructions output at the end of `build_lbann.sh` do not match.  For example, in the docs the following instructions for rebuilding are given:
```
<path to lbann repo>/scripts/build_lbann.sh --build-env-only -- +dihydrogen +cuda +half
cd <path to lbann repo>/spack-build-<hash>
ninja install
```
The first instructions results in an error: `==> Error: [Errno 2] No such file or directory`

The instructions provided at the end of the initial LBANN build with `build_lbann.sh` provide the correct instruction to drop into a shell with the spack build environment, but omit instructions to then run `ninja install` in the build directory.

From what I can tell, the docs need to be updated to agree with the instructions provided in `build_lbann.sh`.

One potential improvement regarding these instructions is to output them to files that the user can easily call (e.g., `./scripts/rebuild_lbann.sh`).  Currently the instructions are quite verbose and if the user does not copy them, they could be easily lost.

**Comment by bvanessen** [comment]
> From trying the new install methods on Pascal...
> 
> USER EXPERIENCE:
> I could not get a non-developer install to work with the instructions provided in the docs (`./scripts/build_lbann.sh -e lbann`), giving the error: `==> Error: invalid values for variant "cuda_arch" in package "lbann": ['60+dihydrogen+distconv~half']`
> 

@mrwyattii Not quite sure what is going on here for you.  When I tried this, I found a bug in the script that didn't properly identify the fact that the environment didn't exist for me.  I have fixed this.  As far as your command, there should be no `cuda_arch` set since you didn't enable the `+cuda` variant.

**Comment by bvanessen** [comment]
> One potential improvement regarding these instructions is to output them to files that the user can easily call (e.g., `./scripts/rebuild_lbann.sh`). Currently the instructions are quite verbose and if the user does not copy them, they could be easily lost.

This is what the script option `./scripts/build_lbann.sh -e lbann --build-env-only bash  -- +cuda` does.  Is that not sufficient.

**Comment by bvanessen** [comment]
@mrwyattii Please run the script again and see if the new output at the end addresses your comment / suggestion for a `rebuild_lbann.sh` script.

**Comment by mrwyattii** [comment]
> 
> 
> > From trying the new install methods on Pascal...
> > USER EXPERIENCE:
> > I could not get a non-developer install to work with the instructions provided in the docs (`./scripts/build_lbann.sh -e lbann`), giving the error: `==> Error: invalid values for variant "cuda_arch" in package "lbann": ['60+dihydrogen+distconv~half']`
> 
> @mrwyattii Not quite sure what is going on here for you. When I tried this, I found a bug in the script that didn't properly identify the fact that the environment didn't exist for me. I have fixed this. As far as your command, there should be no `cuda_arch` set since you didn't enable the `+cuda` variant.

Looking at `$SPACK_ROOT/var/spack/environments/$LBANN_ENV/spack.yaml`, the spec is `lbann@test+cuda cuda_arch=60+dihydrogen+distconv~half`.  I believe this is the line that spack is complaining about for the user install with `./scripts/install_lbann -e $LBANN_ENV`.

The `cuda_arch` is added by `set_center_specific_gpu_arch()` on line 67 of `install_lbann.sh`.

The problem persists after your recent changes

**Comment by mrwyattii** [comment]
> 
> 
> @mrwyattii Please run the script again and see if the new output at the end addresses your comment / suggestion for a `rebuild_lbann.sh` script.

Yes, I think the changes address my original concern.

**Review by mrwyattii — APPROVED** [review]
Recent changes address my earlier concerns, following instructions in the updated docs gives me a working install.  LGTM

**Comment by davidHysom** [comment]
IMHO any sort of build script/system should either 1) print a help message if run without arguments, or 2) "do the right thing" most of the time.

running build_lbann.sh with no arguments gave: "Spack could not activate environment lbann-local-power9le -- install dependencies with -d flag," so that checks of 2).

I then ran build_lbann.sh -d and got this output at the end:
CMake Generate step failed.  Build files cannot be regenerated correctly.
...
  >> 172    CMake Error at docs/CMakeLists.txt:48 (add_custom_target):
     173      Cannot find source file:
     174
     175        /usr/WS2/hysom/lbann/docs/build_spack_extra_config.rst

Additionally: It's not obvious what the new script build_lbann.sh is intended for; is it to replace build_lbann_from_source.sh? Having three build scripts is confusing; there should be a README that tells which one should be used by whom and for what purpose.

I run with a fresh clone of develop, then pulled in the PR.

**Comment by mrwyattii** [comment]
> 
> 
> IMHO any sort of build script/system should either 1) print a help message if run without arguments, or 2) "do the right thing" most of the time.
> 
> running build_lbann.sh with no arguments gave: "Spack could not activate environment lbann-local-power9le -- install dependencies with -d flag," so that checks of 2).
> 
> I then ran build_lbann.sh -d and got this output at the end:
> CMake Generate step failed. Build files cannot be regenerated correctly.
> ...
> 
> > > 172    CMake Error at docs/CMakeLists.txt:48 (add_custom_target):
> > > 173      Cannot find source file:
> > > 174
> > > 175        /usr/WS2/hysom/lbann/docs/build_spack_extra_config.rst
> 
> Additionally: It's not obvious what the new script build_lbann.sh is intended for; is it to replace build_lbann_from_source.sh? Having three build scripts is confusing; there should be a README that tells which one should be used by whom and for what purpose.
> 
> I run with a fresh clone of develop, then pulled in the PR.

I think the build script does 1) and 2).  It informs you to add the `-d` flag and then (for me) it works.  I think you're getting that error because you're trying to build the docs.  `build_spack_extra_config.rst` was removed in this PR and `docs/CMakeLists.txt` should be updated to reflect that and avoid the error you are getting.

Additionally, this PR includes several updates in the docs that instruct the user how to use the new script (see https://github.com/LLNL/lbann/blob/a459ff4054e715bebed8389e2fcc730f91ade323/docs/building_lbann.rst).  My understanding is that the old build scripts will be removed, so that should clear up any confusion about multiple build scripts.

**Comment by timmoon10** [comment]
LeNet runs for me.

**Comment by davidHysom** [comment]
Summary: the build worked, but instructions at the end of running build_lbann.sh (for rebuilding) failed due to incorrect directory specification.
Details; what I ran:

### pascal, compute node, tcsh ###
cd /dev/shm
mdkir hysom
cd hysom
git clone https://github.com/LLNL/lbann.git
cd lbann
git checkout -b bvanessen-feature_reworked_install_process develop
git pull https://github.com/bvanessen/lbann.git feature_reworked_install_process
cd scripts
/usr/bin/time build_lbann.sh -d   |& tee out
...
##########################################################################################
Alternatively, for rebuilding, the script can drop create a shell in the build environment
  ./build_lbann.sh --build-env-only bash -e lbann-local-broadwell --
  cd spack-build-mzsd3vs
  ninja install
##########################################################################################
After running those, the third command gave:
ninja: error: loading 'build.ninja': No such file or directory

$ ls -l /dev/shm/hysom/lbann/scripts/spack-build-mzsd3vs
lrwxrwxrwx 1 hysom hysom 69 Jan 23 06:45 /dev/shm/hysom/lbann/scripts/spack-build-mzsd3vs -> /var/tmp/hysom/lbann-spack-build-mzsd3vs-2021-01-23-064506-1MNQIeBVFE

$ ls -l /var/tmp/hysom/lbann-spack-build-mzsd3vs-2021-01-23-064506-1MNQIeBVFE
total 0

This seems to be where I need to run ninja to rebuild:
$ cd /dev/shm/hysom/lbann/build/pascal.lbann-local-broadwell
$ ls
CMakeCache.txt
...
build.ninja
...

**Comment by bvanessen** [comment]
> Summary: the build worked, but instructions at the end of running build_lbann.sh (for rebuilding) failed due to incorrect directory specification.
> Details; what I ran:
> 
> ### pascal, compute node, tcsh
> cd /dev/shm
> mdkir hysom
> cd hysom
> git clone https://github.com/LLNL/lbann.git
> cd lbann
> git checkout -b bvanessen-feature_reworked_install_process develop
> git pull https://github.com/bvanessen/lbann.git feature_reworked_install_process
> cd scripts
> /usr/bin/time build_lbann.sh -d |& tee out
> ...
> ##########################################################################################
> Alternatively, for rebuilding, the script can drop create a shell in the build environment
> ./build_lbann.sh --build-env-only bash -e lbann-local-broadwell --
> cd spack-build-mzsd3vs
> ninja install
> ##########################################################################################
> After running those, the third command gave:
> ninja: error: loading 'build.ninja': No such file or directory
> 
> $ ls -l /dev/shm/hysom/lbann/scripts/spack-build-mzsd3vs
> lrwxrwxrwx 1 hysom hysom 69 Jan 23 06:45 /dev/shm/hysom/lbann/scripts/spack-build-mzsd3vs -> /var/tmp/hysom/lbann-spack-build-mzsd3vs-2021-01-23-064506-1MNQIeBVFE
> 
> $ ls -l /var/tmp/hysom/lbann-spack-build-mzsd3vs-2021-01-23-064506-1MNQIeBVFE
> total 0
> 
> This seems to be where I need to run ninja to rebuild:
> $ cd /dev/shm/hysom/lbann/build/pascal.lbann-local-broadwell
> $ ls
> CMakeCache.txt
> ...
> build.ninja
> ...

@davidHysom Can you send me more output from your failed build on the PR?  I am trying to figure out why the build directory is wrong.  Also where does /dev/shm/hysom/lbann/build/pascal.lbann-local-broadwellpoint?

**Review by samadejacobs — APPROVED** [review]
WFM

File: scripts/build_lbann.sh [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+echo "  spack build-env ${LBANN_SPEC} -- bash" | tee -a ${LOG}
+echo "  cd spack-build-${LBANN_SPEC_HASH}" | tee -a ${LOG}
+echo "  ninja install" | tee -a ${LOG}
+echo "To use this version of LBANN use the module system without the need for activating the environment (does not require being in an environment)" | tee -a ${LOG}
+echo "  module load lbann/${LBANN_LABEL}-${LBANN_SPEC_HASH}" | tee -a ${LOG}
+echo "or have spack load the module auto-magically. It is installed in a spack environment named ${LBANN_ENV}, access it via: (has to be executed from the environment)"  | tee -a ${LOG}
+echo "  spack load lbann@${LBANN_LABEL} arch=${SPACK_ARCH}" | tee -a ${LOG}
+echo "##########################################################################################" | tee -a ${LOG}
+echo "Alternatively, for rebuilding, the script can drop create a shell in the build environment" | tee -a ${LOG}
+echo "  ${BASH_SOURCE} --build-env-only bash -e ${LBANN_ENV} -- ${CMD_LINE_VARIANTS}" | tee -a ${LOG}
```
Comment (oyamay) [inline scripts/build_lbann.sh:507]: I assume an `-l` option should be used here instead of `-e` as it was removed.

**Review by oyamay — APPROVED** [review]
It worked both with and without the CUDA support in my environment on Lassen.

**Pull request merged**